plugins/x86/x86_backend.ml

```diff
@@ -1,50 +1,61 @@
 open Core_kernel.Std
+open Or_error
 open Bap.Std
 open X86_tools_types
 
 module Insn = Disasm_expert.Basic.Insn
 module Table = String.Table
 
-module type Data = sig
-  val table : lifter Table.t
+type opcode = string
+type prefix = X86_prefix.t
+type pref_lift = mem -> bil -> bil
+
+module type S = sig
+  val register : opcode -> lifter -> unit
+  val handle_prefix : opcode -> prefix -> pref_lift -> unit    
+  module Make (T : Target) : Target
 end
 
-module Make (D : Data) (PR : PR) = struct
-  open D
-  let register op lift =
-    let lift mem insn =
-      try
-        let open Or_error in
-        match PR.parse mem op with
-        | `LOCK_PREFIX -> lift mem insn >>| PR.lock
-        | `REP_PREFIX ->  lift mem insn >>| PR.rep mem
-        | `REPE_PREFIX ->  lift mem insn >>| PR.repe mem
-        | `REPNE_PREFIX -> lift mem insn >>| PR.repne mem
-        | `EXCEPTION no -> Ok [ Bil.cpuexn no ]
-        | `NONE -> lift mem insn
-      with exn -> Or_error.of_exn exn in
-    String.Table.set
-      table
-      ~key:(X86_opcode.sexp_of_t op |> Sexp.to_string)
-      ~data:lift
+module Make (PR : PR) = struct
+
+  type record = {
```

Comment:
Ok, these points sound good, but they actually stand against the current design. If there are prefixes that modify current instruction semantics in a such way, that you can't fix it later on a bil level, then you current design will not work also. Now, the question is, why we're trying to treat prefixes as a separate function. Maybe they should be handled at each particular instruction case? We have a module that provides convenience functions for prefix handling, so probably, instead of registering any callbacks we should just handle the prefixes straightforwardly in each instruction case.